File: src/main/java/net/thauvin/erik/mobibot/modules/CurrencyConverter.java
/*
 * CurrencyConverter.java
 *
 * Copyright (c) 2004-2020, Erik C. Thauvin (erik@thauvin.net)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *   Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *   Neither the name of this project nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package net.thauvin.erik.mobibot.modules;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import net.thauvin.erik.mobibot.Constants;
import net.thauvin.erik.mobibot.Mobibot;
import net.thauvin.erik.mobibot.Utils;
import net.thauvin.erik.mobibot.msg.ErrorMessage;
import net.thauvin.erik.mobibot.msg.Message;
import net.thauvin.erik.mobibot.msg.NoticeMessage;
import net.thauvin.erik.mobibot.msg.PublicMessage;
import org.apache.commons.lang3.StringUtils;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.JDOMException;
import org.jdom2.Namespace;
import org.jdom2.input.SAXBuilder;

import javax.xml.XMLConstants;
import java.io.IOException;
import java.net.URL;
import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

/**
 * The CurrentConverter module.
 *
 * @author <a href="https://erik.thauvin.net" target="_blank">Erik C. Thauvin</a>
 * @created Feb 11, 2004
 * @since 1.0
 */
@SuppressWarnings("PMD.UseConcurrentHashMap")
public final class CurrencyConverter extends ThreadedModule {
    /**
     * The rates keyword.
     */
    static final String CURRENCY_RATES_KEYWORD = "rates";

    // Currency command
    private static final String CURRENCY_CMD = "currency";
    // Exchange rates
    private static final Map<String, String> EXCHANGE_RATES = new TreeMap<>();
    // Exchange rates table URL
    private static final String EXCHANGE_TABLE_URL = "https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml";
    // Last exchange rates table publication date
    private static String pubDate = "";

    /**
     * Creates a new {@link CurrencyConverter} instance.
     */
    public CurrencyConverter() {
        super();
        commands.add(CURRENCY_CMD);
    }

    /**
     * Converts from a currency to another.
     *
     * <p>100 USD to EUR</p>
     *
     * @param query The query.
     * @return The {@link Message} contained the converted currency.
     * @throws ModuleException If an error occurs while converting.
     */
    static Message convertCurrency(final String query) throws ModuleException {
        if (EXCHANGE_RATES.isEmpty()) {
            try {
                final SAXBuilder builder = new SAXBuilder();
                // See https://rules.sonarsourcecom/java/tag/owasp/RSPEC-2755
                builder.setProperty(XMLConstants.ACCESS_EXTERNAL_DTD, "");
                builder.setProperty(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
                builder.setIgnoringElementContentWhitespace(true);

                final Document doc = builder.build(new URL(EXCHANGE_TABLE_URL));
                final Element root = doc.getRootElement();
                final Namespace ns = root.getNamespace("");
                final Element cubeRoot = root.getChild("Cube", ns);
                final Element cubeTime = cubeRoot.getChild("Cube", ns);

                pubDate = cubeTime.getAttribute("time").getValue();

                final List<Element> cubes = cubeTime.getChildren();
                Element cube;

                for (final Element rawCube : cubes) {
                    cube = rawCube;
                    EXCHANGE_RATES.put(
                            cube.getAttribute("currency").getValue(),
                            cube.getAttribute("rate").getValue());
                }

                EXCHANGE_RATES.put("EUR", "1");
            } catch (JDOMException e) {
                throw new ModuleException(query, "An error has occurred while parsing the exchange rates table.", e);
            } catch (IOException e) {
                throw new ModuleException(
                        query, "An error has occurred while fetching the exchange rates table.", e);
            }
        }

        if (EXCHANGE_RATES.isEmpty()) {
            return new ErrorMessage("Sorry, but the exchange rate table is empty.");
        } else {
            final String[] cmds = query.split(" ");

            if (cmds.length == 4) {
                if (cmds[3].equals(cmds[1]) || "0".equals(cmds[0])) {
                    return new ErrorMessage("You're kidding, right?");
                } else {
                    try {
                        final double amt = Double.parseDouble(cmds[0].replace(",", ""));
                        final double from =
                                Double.parseDouble(EXCHANGE_RATES.get(cmds[1].toUpperCase(Constants.LOCALE)));
                        final double to = Double.parseDouble(EXCHANGE_RATES.get(cmds[3].toUpperCase(Constants.LOCALE)));

                        return new PublicMessage(
                                NumberFormat.getCurrencyInstance(Locale.US).format(amt).substring(1)
                                + ' '
                                + cmds[1].toUpperCase(Constants.LOCALE)
                                + " = "
                                + NumberFormat.getCurrencyInstance(Locale.US)
                                              .format((amt * to) / from)
                                              .substring(1)
                                + ' '
                                + cmds[3].toUpperCase(Constants.LOCALE));
                    } catch (Exception e) {
                        throw new ModuleException("convertCurrency(" + query + ')',
                                                  "The supported currencies are: " + EXCHANGE_RATES.keySet(), e);
                    }
                }
            } else if (CURRENCY_RATES_KEYWORD.equals(query)) {

                final StringBuilder buff = new StringBuilder().append('[').append(pubDate).append("]: ");

                int i = 0;
                for (final Map.Entry<String, String> rate : EXCHANGE_RATES.entrySet()) {
                    if (i > 0) {
                        buff.append(", ");
                    }
                    buff.append(rate.getKey()).append(": ").append(rate.getValue());
                    i++;
                }

                return new NoticeMessage(buff.toString());
            }
        }
        return new ErrorMessage("The supported currencies are: " + EXCHANGE_RATES.keySet());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void commandResponse(final Mobibot bot,
                                final String sender,
                                final String cmd,
                                final String args,
                                final boolean isPrivate) {
        synchronized (this) {
            if (!pubDate.equals(Utils.today())) {
                EXCHANGE_RATES.clear();
            }
        }

        super.commandResponse(bot, sender, cmd, args, isPrivate);
    }

    /**
     * Converts the specified currencies.
     */
    @SuppressFBWarnings("REDOS")
    @Override
    void run(final Mobibot bot, final String sender, final String cmd, final String query) {
        if (StringUtils.isNotBlank(sender) && StringUtils.isNotBlank(query)) {
            if (query.matches("\\d+([,\\d]+)?(\\.\\d+)? [a-zA-Z]{3}+ to [a-zA-Z]{3}+")) {
                try {
                    final Message msg = convertCurrency(query);
                    if (msg.isError()) {
                        helpResponse(bot, sender, CURRENCY_CMD + ' ' + query, false);
                    }
                    bot.send(sender, msg);
                } catch (ModuleException e) {
                    bot.getLogger().warn(e.getDebugMessage(), e);
                    bot.send(sender, e.getMessage());
                }
            } else {
                helpResponse(bot, sender, CURRENCY_CMD + ' ' + query, true);
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void helpResponse(final Mobibot bot, final String sender, final String args, final boolean isPrivate) {
        bot.send(sender, "To convert from one currency to another:");
        bot.send(sender, bot.helpIndent(bot.getNick() + ": " + CURRENCY_CMD + " [100 USD to EUR]"));

        if (args.endsWith(CURRENCY_CMD)) {
            bot.send(sender, "For a listing of currency rates:");
            bot.send(sender, bot.helpIndent(bot.getNick() + ": " + CURRENCY_CMD) + ' ' + CURRENCY_RATES_KEYWORD);
            bot.send(sender, "For a listing of supported currencies:");
            bot.send(sender, bot.helpIndent(bot.getNick() + ": " + CURRENCY_CMD));
        }
    }
}

File: src/main/kotlin/net/thauvin/erik/mobibot/modules/CurrencyConverter.kt
/*
 * CurrencyConverter.kt
 *
 * Copyright 2004-2025 Erik C. Thauvin (erik@thauvin.net)
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *   Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *   Neither the name of this project nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package net.thauvin.erik.mobibot.modules

import net.thauvin.erik.mobibot.Utils.bot
import net.thauvin.erik.mobibot.Utils.helpCmdSyntax
import net.thauvin.erik.mobibot.Utils.helpFormat
import net.thauvin.erik.mobibot.Utils.reader
import net.thauvin.erik.mobibot.Utils.sendList
import net.thauvin.erik.mobibot.Utils.sendMessage
import net.thauvin.erik.mobibot.msg.ErrorMessage
import net.thauvin.erik.mobibot.msg.Message
import net.thauvin.erik.mobibot.msg.PublicMessage
import org.json.JSONObject
import org.pircbotx.hooks.types.GenericMessageEvent
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URL
import java.text.DecimalFormat
import java.util.*


/**
 * Converts between currencies.
 */
class CurrencyConverter : AbstractModule() {
    override val name = "CurrencyConverter"

    companion object {
        /**
         * The API Key property.
         */
        const val API_KEY_PROP = "exchangerate-api-key"

        // Currency command
        private const val CURRENCY_CMD = "currency"

        // Currency codes keyword
        private const val CODES_KEYWORD = "codes"

        // Decimal format
        private val DECIMAL_FORMAT = DecimalFormat("0.00#")

        // Empty symbols table.
        private const val EMPTY_SYMBOLS_TABLE = "Sorry, but the currency table is empty."

        // Logger
        private val LOGGER: Logger = LoggerFactory.getLogger(CurrencyConverter::class.java)

        // Currency symbols
        private val SYMBOLS: TreeMap<String, String> = TreeMap()

        /**
         * Converts from a currency to another.
         */
        @JvmStatic
        fun convertCurrency(apiKey: String?, query: String): Message {
            if (apiKey.isNullOrEmpty()) {
                throw ModuleException("${CURRENCY_CMD}($query)", "No Exchange Rate API key specified.")
            }

            val cmds = query.split(" ")
            return if (cmds.size == 4) {
                if (cmds[3] == cmds[1] || "0" == cmds[0]) {
                    PublicMessage("You're kidding, right?")
                } else {
                    val to = cmds[1].uppercase()
                    val from = cmds[3].uppercase()
                    if (SYMBOLS.contains(to) && SYMBOLS.contains(from)) {
                        try {
                            val amt = cmds[0].replace(",", "")
                            val url = URL("https://v6.exchangerate-api.com/v6/$apiKey/pair/$to/$from/$amt")
                            val body = url.reader().body
                            val json = JSONObject(body)

                            if (json.getString("result") == "success") {
                                val result = DECIMAL_FORMAT.format(json.getDouble("conversion_result"))
                                PublicMessage(
                                    "${cmds[0]} ${SYMBOLS[to]} = $result ${SYMBOLS[from]}"
                                )
                            } else {
                                ErrorMessage("Sorry, an error occurred while converting the currencies.")
                            }
                        } catch (ioe: IOException) {
                            if (LOGGER.isWarnEnabled) {
                                LOGGER.warn("IO error while converting currencies: ${ioe.message}", ioe)
                            }
                            ErrorMessage("Sorry, an IO error occurred while converting the currencies.")
                        }
                    } else {
                        ErrorMessage("Sounds like monopoly money to me!")
                    }
                }
            } else {
                ErrorMessage("Invalid query. Let's try again.")
            }
        }

        /**
         * Loads the currency ISO symbols.
         */
        @JvmStatic
        @Throws(ModuleException::class)
        fun loadSymbols(apiKey: String?) {
            if (!apiKey.isNullOrEmpty()) {
                try {
                    val url = URL("https://v6.exchangerate-api.com/v6/$apiKey/codes")
                    val json = JSONObject(url.reader().body)
                    if (json.getString("result") == "success") {
                        val codes = json.getJSONArray("supported_codes")
                        for (i in 0 until codes.length()) {
                            val code = codes.getJSONArray(i)
                            SYMBOLS[code.getString(0)] = code.getString(1)
                        }
                    }
                } catch (e: IOException) {
                    throw ModuleException(
                        "loadCodes(): IOE",
                        "An IO error has occurred while retrieving the currencies.",
                        e
                    )
                }
            }
        }
    }

    init {
        commands.add(CURRENCY_CMD)
        initProperties(API_KEY_PROP)
        loadSymbols(properties[ChatGpt2.API_KEY_PROP])
    }

    // Reload currency codes
    private fun reload(apiKey: String?) {
        if (!apiKey.isNullOrEmpty() && SYMBOLS.isEmpty()) {
            try {
                loadSymbols(apiKey)
            } catch (e: ModuleException) {
                if (LOGGER.isWarnEnabled) LOGGER.warn(e.debugMessage, e)
            }
        }
    }

    /**
     * Converts the specified currencies.
     */
    override fun commandResponse(channel: String, cmd: String, args: String, event: GenericMessageEvent) {
        reload(properties[API_KEY_PROP])

        when {
            SYMBOLS.isEmpty() -> {
                event.respond(EMPTY_SYMBOLS_TABLE)
            }

            args.matches("\\d+([,\\d]+)?(\\.\\d+)? [a-zA-Z]{3}+ (to|in) [a-zA-Z]{3}+".toRegex()) -> {
                val msg = convertCurrency(properties[API_KEY_PROP], args)
                event.respond(msg.msg)
                if (msg.isError) {
                    helpResponse(event)
                }
            }

            args.contains(CODES_KEYWORD) -> {
                event.sendMessage("The supported currency codes are:")
                event.sendList(SYMBOLS.keys.toList(), 11, isIndent = true)
            }

            else -> {
                helpResponse(event)
            }
        }
    }

    override fun helpResponse(event: GenericMessageEvent): Boolean {
        reload(properties[API_KEY_PROP])

        if (SYMBOLS.isEmpty()) {
            event.sendMessage(EMPTY_SYMBOLS_TABLE)
        } else {
            val nick = event.bot().nick
            event.sendMessage("To convert from one currency to another:")
            event.sendMessage(helpFormat(helpCmdSyntax("%c $CURRENCY_CMD 100 USD to EUR", nick, isPrivateMsgEnabled)))
            event.sendMessage(
                helpFormat(
                    helpCmdSyntax("%c $CURRENCY_CMD 50,000 GBP to USD", nick, isPrivateMsgEnabled)
                )
            )
            event.sendMessage("To list the supported currency codes:")
            event.sendMessage(
                helpFormat(
                    helpCmdSyntax("%c $CURRENCY_CMD $CODES_KEYWORD", nick, isPrivateMsgEnabled)
                )
            )
        }
        return true
    }
}
